### PR TITLE
enable loading older TE checkpoints

### PR DESCRIPTION
**Pull request by dimapihtar**
# What does this PR do ?

Enables loading older TE checkpoints with avoiding "extra state" layer missmatch error.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 1]
```diff
@@ -1998,7 +1999,8 @@ def on_load_checkpoint(self, checkpoint) -> None:
                         key.replace('model.', ''): checkpoint_state_dict.pop(key)
                         for key in list(checkpoint_state_dict.keys())
                     }
-                    module.load_state_dict(checkpoint_state_dict, strict=True)
+                    strict = False if self.load_legacy else True
```
Comment (janekl) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:2002]: NIT: `strict = not self.load_legacy`

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 2]
```diff
@@ -1998,7 +1998,16 @@ def on_load_checkpoint(self, checkpoint) -> None:
                         key.replace('model.', ''): checkpoint_state_dict.pop(key)
                         for key in list(checkpoint_state_dict.keys())
                     }
-                    module.load_state_dict(checkpoint_state_dict, strict=True)
+                    try:
+                        module.load_state_dict(checkpoint_state_dict, strict=True)
+                    except Exception as e:
```
Comment (janekl) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:2003]: We could consider explicitly catching just `RuntimeError` that is actually thrown here for missing key(s).

Currently we'll retry on any problem while we just expect to have a workaround for that keys error

Comment (dimapihtar) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:2003]: @janekl, you're right, thanks. I've just forgot to change it before pushing.

**Review by janekl — DISMISSED** [review]

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 3]
```diff
@@ -1998,7 +1998,16 @@ def on_load_checkpoint(self, checkpoint) -> None:
...
                         for key in list(checkpoint_state_dict.keys())
                     }
-                    module.load_state_dict(checkpoint_state_dict, strict=True)
+                    try:
+                        module.load_state_dict(checkpoint_state_dict, strict=True)
+                    except RuntimeError as e:
+                        missing_keys, expected_keys = module.load_state_dict(checkpoint_state_dict, strict=False)
+                        if all(s.endswith('_extra_state') for s in missing_keys):
+                            logging.warning(
+                                f'Loding checkpoint with old Transformer Engine version. Missing layers {missing_keys} will be ignored.'
```
Comment (mikolajblaz) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:2007]: Loding -> Loading

`with old Transformer Engine version` -> `created with Transformer Engine lower than 1.13`

**Review by thomasdhc — APPROVED** [review]
LGTM
